src/providers/azure.rs

```diff
@@ -0,0 +1,503 @@
+mod azcmd {
+    use super::Error;
+    use super::Region;
+    use failure::ResultExt;
+    use serde::{Deserialize, Serialize};
+    use std::process::Command;
+
+    pub fn available_instances_in_region(r: Region) -> Result<Vec<String>, Error> {
+        #[allow(non_snake_case)]
+        #[derive(Debug, Deserialize, Serialize)]
+        struct VmSizeDescriptor {
+            maxDataDiskCount: usize,
+            memoryInMb: usize,
+            name: String,
+            numberOfCores: usize,
+            osDiskSizeInMb: usize,
+            resourceDiskSizeInMb: usize,
+        }
+
+        let out = Command::new("az")
+            .args(&["vm", "list-sizes", "-l", &r.to_string()])
+            .output()?;
+        if !out.status.success() {
+            bail!("Failed to get available instance sizes");
+        }
+
+        let v: Vec<VmSizeDescriptor> = serde_json::from_slice(&out.stdout)?;
+        Ok(v.into_iter().map(|x| x.name).collect())
+    }
+
+    pub fn create_resource_group(r: Region, name: &str) -> Result<(), Error> {
+        let out = Command::new("az")
+            .args(&[
+                "group",
+                "create",
+                "--name",
+                name,
+                "--location",
+                &r.to_string(),
+            ])
+            .output()?;
+
+        if !out.status.success() {
+            bail!("Failed to create resource group {} in region {:?}", name, r)
+        }
+
+        Ok(())
+    }
+
+    pub fn create_vm(rg: &str, name: &str, size: &str, username: &str) -> Result<String, Error> {
+        #[allow(non_snake_case)]
+        #[derive(Debug, Deserialize, Serialize)]
+        struct VmCreateOut {
+            powerState: String,
+            publicIpAddress: String,
+            resourceGroup: String,
+        }
+
+        let out = Command::new("az")
+            .args(&[
+                "vm",
+                "create",
+                "--resource-group",
+                rg,
+                "--name",
+                name,
+                "--image",
+                "UbuntuLTS",
+                "--size",
+                size,
+                "--admin-username",
+                username,
+                "--ssh-key-value",
+                "@~/.ssh/id_rsa.pub",
+            ])
+            .output()?;
+
+        if !out.status.success() {
+            return Err(format_err!("Failed to create vm {}", name))
+                .context(String::from_utf8(out.stderr).unwrap())?;
+        }
+
+        let vm: VmCreateOut = serde_json::from_slice(&out.stdout)?;
+        ensure!(vm.powerState == "VM running", "VM power state incorrect");
+        ensure!(vm.resourceGroup == rg, "VM resource group incorrect");
+        Ok(vm.publicIpAddress)
+    }
+
+    pub fn open_ports(rg: &str, vm_name: &str) -> Result<(), Error> {
+        let out = Command::new("az")
+            .args(&[
+                "vm",
+                "open-port",
+                "--port",
+                "0-65535",
```

Comment:
This seems too broad. It should probably only open for traffic *between* the VMs, not all ports externally?

Reply:
The right answer is probably to make which ports to open configurable, but still allow any inbound IPs - my use cases, for example, often have communication from "outside" machines.

Reply:
I think at the very least we should make sure the AWS and Azure configurations match up.

Reply:
They do match up: https://github.com/akshayknarayan/tsunami/blob/machine-lifetimes/src/providers/aws.rs#L397

Reply:
Ah, interesting. I feel like my original intention there was to open up only VM-internal communication, but that would rely on VPC, which is still left as a `TODO` there. I guess this is fine for now!